Design/algorithm_refactoring.md

```diff
@@ -0,0 +1,209 @@
+# Cleaning up Mantid Algorithms
+
+## Motivation
+
+See discussions during the developer workshop in January 2016.
+
+This document is meant as a base for discussion and does not (yet) represent a complete design. The results of this and potentially various parts of this document will eventually be used to create documentation for developers on how to write an algorithm.
+
+
+## N kinds of boilerplate code
+
+
+### Progress
+
+```cpp
+Progress progress(this, 0.0, 1.0, m_numberOfSpectra);
+// [...]
+progress.report();
+```
+
+This can be done in the proposed `Algorithm::transform` method, see below.
+
+
+### Getters on workspaces
+
+Typically when looping over spectra, one or several getters are used on workspaces, for event lists, spectra, and histograms. For example, `getEventList()`, `getSpectrum()`, and `readX()`, `readY()`, `readE()`.
+
+There may be a way to combine this with what is described in the proposed `Algorithm::transform` method below, but it is not obvious how to get rid of these entirely.
+
+
+### Looping over spectra
+
+The loop over histograms comprises various kinds of boilerplate:
+
+- The `for`-loop statement itself
+- `PARALLEL` macros, which in turn contain:
+  - OpenMP pragmas
+  - A mechanism for cancellation
+  - Parallel exception handling
+
+  Typically there are 4 `PARALLEL` macros for a single loop over spectra.
+
+To get rid of these, we would like to implement something along the lines of `std::transform`. In our case it turns out to be most convenient to make this a member of `Algorithm`, since our `transform` needs access to various members of `Algorithm`, in particular:
+
+- The `PARALLEL` macros access `m_cancel`, `m_parallelException` (which, however, could be made local), and `g_log` (which may not be necessary, strictly speaking).
+- We would like to move the `Progress` code into our `transform`, so we need the `this` pointer.
+
+If it were not a member we would need some way of passing all these things into our `transform`, which is cumbersome.
+
+In the simplest case (in-place modification of a single field), this could be implemented along the lines of this:
+
+```cpp
+// Note that we need the explicit syntax ELEM (WS::*getter)(IDX), else the compiler cannot resolve (const) overloads
+template <class WS, class OP, class ELEM, class IDX>
```

Comment:
I think this is a good plan. Some algorithms, like `DownloadInstrument` certainly doesn't want any of this stuff. It would be good to try and make the `Algorithm` class simpler and move things out.